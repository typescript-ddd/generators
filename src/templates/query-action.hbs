import { QueryBus, QueryAction } from "@typescript-ddd/core";
import { {{queryName}}Query } from "../../infrastructure";

/**
 * Represents {{noun actionName}} query action. 
 */
export class {{pascal actionName}}QueryAction implements QueryAction {
  method: string = "GET";
  path: string = "{{path}}";

  /**
   * Initializes a new instance of {{noun actionName}} query action.
   * @param {QueryBus} queryBus - The query bus.
   */
  constructor(private readonly queryBus: QueryBus) {}

  /**
   * Executes {{noun actionName}} query action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    {{#if requestParams}}
    {{#each requestParams}}
    const {{name}} = req.params.{{name}};
    {{/each}}
    const value = await this.queryBus.ask(new {{queryName}}Query(
      {{#each requestParams}}
      {{name}},
      {{/each}}
    ));
    {{else}}
    const value = await this.queryBus.ask(new {{queryName}}Query());
    {{/if}}
    res.status(200).json(value);
  }
}
