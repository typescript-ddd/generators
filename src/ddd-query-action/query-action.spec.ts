import { QueryActionTemplateData } from "./query-action.types";
import { generateQueryAction } from "./query-action.generator";

describe("QueryAction", () => {

  it("should render template", () => {
    // Arrange
    const data: QueryActionTemplateData = {
      path: "/users",
      actionName: "GetUsers",
      queryName: "FindUsers",
    };

    // Act
    const result = generateQueryAction(data);

    // Assert
    expect(result).toMatchSnapshot();
  });

  it("should render template with request params", () => {
    // Arrange
    const data: QueryActionTemplateData = {
      path: "/user/:id",
      actionName: "GetUser",
      queryName: "FindUser",
      requestParams: [
        { name: "id", type: "string" },
      ]
    };

    // Act
    const result = generateQueryAction(data);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
