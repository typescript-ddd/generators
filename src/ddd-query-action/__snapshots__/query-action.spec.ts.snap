// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryAction should render template 1`] = `
"import { QueryBus, QueryAction } from "@typescript-ddd/core";
import { FindUsersQuery } from "../../infrastructure";

/**
 * Represents a get user query action. 
 */
export class GetUsersQueryAction implements QueryAction {
  method: string = "GET";
  path: string = "/users";

  /**
   * Initializes a new instance of a get user query action.
   * @param {QueryBus} queryBus - The query bus.
   */
  constructor(private readonly queryBus: QueryBus) {}

  /**
   * Executes a get user query action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    const value = await this.queryBus.ask(new FindUsersQuery());
    res.status(200).json(value);
  }
}
"
`;

exports[`QueryAction should render template with request params 1`] = `
"import { QueryBus, QueryAction } from "@typescript-ddd/core";
import { FindUserQuery } from "../../infrastructure";

/**
 * Represents a get user query action. 
 */
export class GetUserQueryAction implements QueryAction {
  method: string = "GET";
  path: string = "/user/:id";

  /**
   * Initializes a new instance of a get user query action.
   * @param {QueryBus} queryBus - The query bus.
   */
  constructor(private readonly queryBus: QueryBus) {}

  /**
   * Executes a get user query action.
   * @returns {Promise<void>}
   */
  public async execute(req: any, res: any): Promise<void> {
    const id = req.params.id;
    const value = await this.queryBus.ask(new FindUserQuery(
      id,
    ));
    res.status(200).json(value);
  }
}
"
`;
